### ISSUE TITLE
Memory leak with ranges argument

### ISSUE THREAD

**Issue by antoinediez** [getkeops/keops#342]
Dear KeOps developers,

I have noticed a strange and unfortunate memory leak which happens when the `ranges` argument of a reduction kernel is given. The following MWE is adapted from the [Block-Sparse-Reduction tutorial](https://www.kernel-operations.io/keops/_auto_examples/pytorch/plot_grid_cluster_pytorch.html). The output shown is obtained on a fresh Google Colab session. 

**Note**: the problem seems related to #284 (but the fix by @bcharlier does not seem to solve the issue here) and to #314 (but here I don't see any CPU/GPU mismatch as mentioned by @NightWinkle).

------------------------------------------

Standard imports: 

```python
import numpy as np
import torch
from pykeops.torch import LazyTensor
from pykeops.torch.cluster import grid_cluster
from pykeops.torch.cluster import cluster_ranges_centroids
from pykeops.torch.cluster import sort_clusters
from pykeops.torch.cluster import from_matrix
import nvsmi

use_cuda = torch.cuda.is_available()
if use_cuda:
    torch.set_default_tensor_type("torch.cuda.FloatTensor")   #All tensors should be on the GPU
    device = "cuda"
    
M, N = (100000, 100000)    #Dimensions
T = 5000   #Number of iterations
```

The following function computes a simple reduction and is directly copied from the tutorial. The `ranges` argument of the reduction kernel is updated when the argument `update_ranges` is `True` (so the leak comes from this line and not from the variables computed in the previous ones). 

```python
def bsr_test(update_ranges=True):
    t = torch.linspace(0, 2 * np.pi, M + 1)[:-1]
    x = torch.stack((0.4 + 0.4 * (t / 7) * t.cos(), 0.5 + 0.3 * t.sin()), 1)
    x = x + 0.01 * torch.randn(x.shape)

    y = torch.randn((N,2))
    y = y / 10 + torch.tensor([0.6, 0.6])

    eps = 0.05  # Size of our square bins
    x_labels = grid_cluster(x, eps)  # class labels
    y_labels = grid_cluster(y, eps)  # class labels
    torch.cuda.synchronize()

    x_ranges, x_centroids, _ = cluster_ranges_centroids(x, x_labels)
    y_ranges, y_centroids, _ = cluster_ranges_centroids(y, y_labels)
    torch.cuda.synchronize()

    x, x_labels = sort_clusters(x, x_labels)
    y, y_labels = sort_clusters(y, y_labels)
    torch.cuda.synchronize()

    sigma = 0.05  # Characteristic length of interaction
    # Compute a coarse Boolean mask:
    D = ((x_centroids[:, None, :] - y_centroids[None, :, :]) ** 2).sum(2)
    keep = D < (4 * sigma) ** 2

    ranges_ij = from_matrix(x_ranges, y_ranges, keep)
    torch.cuda.synchronize()

    x_, y_ = x / sigma, y / sigma
    x_i, y_j = LazyTensor(x_[:, None, :]), LazyTensor(y_[None, :, :])
    D_ij = ((x_i - y_j) ** 2).sum(dim=2)  # Symbolic (M,N,1) matrix of squared distances
    K = (-D_ij / 2).exp()  # Symbolic (M,N,1) Gaussian kernel matrix

    b = torch.randn((N, 1))
    
    if update_ranges:
        K.ranges = ranges_ij
    
    a_sparse = K @ b
    
    return a_sparse

```

When the the `ranges` argument is given, there is a small but steady memory leak. 

```python
for it in range(T):
    a = bsr_test(update_ranges=True)
    if (it+1)%1000==0:
        print("Iteration: " + str(it),flush=True)
        print("Memory allocated: " + str(torch.cuda.memory_allocated()),flush=True)
        print("Memory cached: " + str(torch.cuda.memory_reserved()),flush=True)
        print(nvsmi.get_gpu_processes()[0],flush=True)
        print("------",flush=True)
```

Output:
```
Iteration: 999
Memory allocated: 800768
Memory cached: 33554432
pid: 112537 | gpu_id: 0 | gpu_uuid: GPU-d764ada8-9783-d1e3-a88e-8dc94ff22bda | gpu_name: Tesla T4 | used_memory:   670.0MB
------
Iteration: 1999
Memory allocated: 800768
Memory cached: 33554432
pid: 112537 | gpu_id: 0 | gpu_uuid: GPU-d764ada8-9783-d1e3-a88e-8dc94ff22bda | gpu_name: Tesla T4 | used_memory:   676.0MB
------
Iteration: 2999
Memory allocated: 800768
Memory cached: 33554432
pid: 112537 | gpu_id: 0 | gpu_uuid: GPU-d764ada8-9783-d1e3-a88e-8dc94ff22bda | gpu_name: Tesla T4 | used_memory:   682.0MB
------
Iteration: 3999
Memory allocated: 800768
Memory cached: 33554432
pid: 112537 | gpu_id: 0 | gpu_uuid: GPU-d764ada8-9783-d1e3-a88e-8dc94ff22bda | gpu_name: Tesla T4 | used_memory:   688.0MB
------
Iteration: 4999
Memory allocated: 800768
Memory cached: 33554432
pid: 112537 | gpu_id: 0 | gpu_uuid: GPU-d764ada8-9783-d1e3-a88e-8dc94ff22bda | gpu_name: Tesla T4 | used_memory:   698.0MB
------
```
The used memory grows from 670MB to 698MB. This does not happen when the `ranges` argument is not specified. 

```python
for it in range(T):
    a = bsr_test(update_ranges=False)
    if (it+1)%1000==0:
        print("Iteration: " + str(it),flush=True)
        print("Memory allocated: " + str(torch.cuda.memory_allocated()),flush=True)
        print("Memory cached: " + str(torch.cuda.memory_reserved()),flush=True)
        print(nvsmi.get_gpu_processes()[0],flush=True)
        print("------",flush=True)
```

Output:

```
Iteration: 999
Memory allocated: 800768
Memory cached: 33554432
pid: 164027 | gpu_id: 0 | gpu_uuid: GPU-d764ada8-9783-d1e3-a88e-8dc94ff22bda | gpu_name: Tesla T4 | used_memory:   664.0MB
------
Iteration: 1999
Memory allocated: 800768
Memory cached: 33554432
pid: 164027 | gpu_id: 0 | gpu_uuid: GPU-d764ada8-9783-d1e3-a88e-8dc94ff22bda | gpu_name: Tesla T4 | used_memory:   664.0MB
------
Iteration: 2999
Memory allocated: 800768
Memory cached: 33554432
pid: 164027 | gpu_id: 0 | gpu_uuid: GPU-d764ada8-9783-d1e3-a88e-8dc94ff22bda | gpu_name: Tesla T4 | used_memory:   664.0MB
------
Iteration: 3999
Memory allocated: 800768
Memory cached: 33554432
pid: 164027 | gpu_id: 0 | gpu_uuid: GPU-d764ada8-9783-d1e3-a88e-8dc94ff22bda | gpu_name: Tesla T4 | used_memory:   664.0MB
------
Iteration: 4999
Memory allocated: 800768
Memory cached: 33554432
pid: 164027 | gpu_id: 0 | gpu_uuid: GPU-d764ada8-9783-d1e3-a88e-8dc94ff22bda | gpu_name: Tesla T4 | used_memory:   664.0MB
------
```

**Comment by jeanfeydy** [comment]
Hi @antoinediez,

Thanks for your detailed report, and apologies to @NightWinkle for having missed the newest comments on the closed thread. This is an issue that has already bothered several users (especially in a deep learning context), but without clean minimal buggy examples to pinpoint the issue. I think that we have a good enough lead to solve the problem now. I'm very busy this week, but will have a look next week for my next coding session. That's definitely a top priority bug.

See you soon,
Jean

**Comment by antoinediez** [comment]
Hi Jean! 

Thanks, no problem this is not really urgent. 

I know that many things have changed since then but if it can help, I am quite sure that there was no memory leak in the earlier versions of KeOps (certainly before KeOps 2.0 maybe after). 

Best, 
Antoien

**Comment by joanglaunes** [comment]
Hello @antoinediez ,
This is fixed now. I looked at this because I had recently cleaned some stuff in the way we allocate and free memory in the Cuda part of the code. But finally the bug was not related and it was not truely a memory leak. There was a small error in the python part that caused a new instance of the same cuda module to be created at each call (so each iteration), instead of being reused, and these instances are only deleted a the end of the session.
